firebase-vertexai/src/main/kotlin/com/google/firebase/vertexai/type/LiveSession.kt

```diff
@@ -16,359 +16,474 @@
   }
 
-  private suspend fun sendAudioDataToServer() {
+  /**
+   * Sends [data][Content] to the model.
+   *
+   * Calling this after [startAudioConversation] will play the response audio immediately.
+   *
+   * @param content Client [Content] to be sent to the model.
+   */
+  public suspend fun send(content: Content) {
+    val jsonString =
+      Json.encodeToString(LiveClientContentSetup(listOf(content.toInternal()), true).toInternal())
+    session.send(Frame.Text(jsonString))
+  }
 
-    val audioBufferStream = ByteArrayOutputStream()
-    while (isRecording) {
-      val receivedAudio = audioQueue.poll() ?: continue
-      audioBufferStream.write(receivedAudio)
-      if (audioBufferStream.size() >= MIN_BUFFER_SIZE) {
-        sendMediaStream(listOf(MediaData(audioBufferStream.toByteArray(), "audio/pcm")))
-        audioBufferStream.reset()
-      }
-    }
+  /**
+   * Sends text to the model.
+   *
+   * Calling this after [startAudioConversation] will play the response audio immediately.
+   *
+   * @param text Text to be sent to the model.
+   */
+  public suspend fun send(text: String) {
+    send(Content.Builder().text(text).build())
   }
 
-  private fun fillServerResponseAudioQueue(
-    functionCallsHandler: ((FunctionCallPart) -> FunctionResponsePart)? = null
-  ) {
-    CoroutineScope(backgroundDispatcher).launch {
-      receive().collect {
-        if (!isRecording) {
-          cancel()
-        }
-        when (it.status) {
-          LiveContentResponse.Status.INTERRUPTED ->
-            while (!playBackQueue.isEmpty()) playBackQueue.poll()
-          LiveContentResponse.Status.NORMAL ->
-            if (!it.functionCalls.isNullOrEmpty() && functionCallsHandler != null) {
-              sendFunctionResponse(it.functionCalls.map(functionCallsHandler).toList())
-            } else {
-              val audioData = it.data?.parts?.get(0)?.asInlineDataPartOrNull()?.inlineData
-              if (audioData != null) {
-                playBackQueue.add(audioData)
-              }
-            }
-        }
-      }
-    }
+  /**
+   * Closes the client session.
+   *
+   * Once a [LiveSession] is closed, it can not be reopened; you'll need to start a new
+   * [LiveSession].
+   *
+   * @see stopReceiving
+   */
+  public suspend fun close() {
+    session.close()
+    stopAudioConversation()
   }
 
-  private fun playServerResponseAudio() {
-    CoroutineScope(backgroundDispatcher).launch {
-      while (isRecording) {
-        val data = playBackQueue.poll()
-        if (data == null) {
-          audioHelper?.start()
-          continue
-        }
-        audioHelper?.stopRecording()
-        audioHelper?.playAudio(data)
-      }
-    }
+  /** Listen to the user's microphone and send the data to the model. */
+  private fun recordUserAudio() {
+    // Buffer the recording so we can keep recording while data is sent to the server
+    audioHelper
+      ?.listenToRecording()
+      ?.buffer(UNLIMITED)
```

Comment:
Is this necessary?

Reply:
Yes, it avoids the recorder's internal buffer from overflowing while we process and send the data to the model.